private/bufpkg/bufconnect/interceptors.go

```diff
@@ -82,25 +71,66 @@ func NewAuthorizationInterceptorProvider(container app.EnvContainer) func(string
 	}
 }
 
-// NewAuthorizationInterceptorProviderWithToken returns a new provider function which, when invoked, returns an
-// interceptor which sets the provided auth token into the request header.  This is used for registry providers where
-// the token is known at provider creation (i.e. when logging in and explicitly pasting a token into stdin
-//
-// Note that the interceptor returned from this provider is always applied LAST in the series of interceptors added to
-// a client.
-func NewAuthorizationInterceptorProviderWithToken(token string) func(string) connect.UnaryInterceptorFunc {
-	return func(address string) connect.UnaryInterceptorFunc {
-		interceptor := func(next connect.UnaryFunc) connect.UnaryFunc {
-			return connect.UnaryFunc(func(
-				ctx context.Context,
-				req connect.AnyRequest,
-			) (connect.AnyResponse, error) {
-				if token != "" {
-					req.Header().Set(AuthenticationHeader, AuthenticationTokenPrefix+token)
-				}
-				return next(ctx, req)
-			})
+// TokenSet is used to provide authentication token in NewAuthorizationInterceptorProvider
+type TokenSet struct {
+	envKey          string
+	bufToken        string
+	remoteUsernames map[string]string
+	remoteTokens    map[string]string
+}
+
+// NewTokenSetFromContainer creates a TokenSet from the BUF_TOKEN environment variable
+func NewTokenSetFromContainer(container app.EnvContainer) (*TokenSet, error) {
+	bufToken := container.Env(tokenEnvKey)
+	tokenSet, err := NewTokenSetFromString(bufToken)
+	if err != nil {
+		return nil, err
+	}
+	tokenSet.envKey = tokenEnvKey
+	return tokenSet, nil
+}
+
+// NewTokenSetFromString creates a TokenSet by the token provided
+func NewTokenSetFromString(token string) (*TokenSet, error) {
+	tokenSet := &TokenSet{
+		remoteUsernames: make(map[string]string),
+		remoteTokens:    make(map[string]string),
+	}
+	tokens := strings.Split(token, ",")
+	for _, u := range tokens {
+		if contain := strings.ContainsAny(u, "@"); contain {
+			keyPairsAndRemoteAddress := strings.Split(u, "@")
+			if len(keyPairsAndRemoteAddress) != 2 {
+				return nil, fmt.Errorf("cannot parse token: %s, invalid remote token: %s", token, u)
+			}
+			keyPairs := strings.Split(keyPairsAndRemoteAddress[0], ":")
+			if len(keyPairs) != 2 {
+				return nil, fmt.Errorf("cannot parse token: %s, invalid remote token: %s", token, u)
+			}
+			remoteAddress := keyPairsAndRemoteAddress[1]
+			username := keyPairs[0]
+			remoteToken := keyPairs[1]
```

Comment:
Try using strings.Cut, like in my comment above.

Do read rsc's "The Unreasonable Effectiveness of Cut" `https://github.com/golang/go/issues/46336` (escaped to avoid pinging that issue). It's a goodie.

Reply:
This code can look way more elegant if we are using `strings.Cut`. But detecting invalid tokens like `invalid:invalid:invalid@remote` or invalid:invalid@invalid@remote is harder. But I think we don't have to valid date the token here as we are only trying to parse the tokens. The tokens are validated in a later step. Should I trying using `strings.Cut` and pass in whatever is parsed?